lib/handshake_common.ml

```diff
@@ -248,21 +243,27 @@ let validate_chain authenticator certificates hostname keytype usage =
     in
     guard (List.for_all check cs) Packet.INSUFFICIENT_SECURITY
 
+  and parse_certificates certs =
+    let certificates = filter_map ~f:parse certs in
+    guard (List.length certs = List.length certificates) Packet.BAD_CERTIFICATE >|= fun () ->
+    certificates
+
   in
 
   (* RFC5246: must be x509v3, take signaturealgorithms into account! *)
   (* RFC2246/4346: is generally x509v3, signing algorithm for certificate _must_ be same as algorithm for certificate key *)
-
+  parse_certificates certificates >>= fun certs ->
   match authenticator with
   | None ->
-    parse certificates >|= fun (s, xs) ->
-    ((s, xs), None)
+    return (certs, None)
   | Some authenticator ->
-    parse certificates >>= fun (s, xs) ->
-    key_size Config.min_rsa_key_size (s :: xs) >>= fun () ->
-    authenticate authenticator hostname (s, xs) >>= fun anchor ->
-    guard (validate_keytype s keytype &&
-           validate_usage s usage &&
-           validate_ext_usage s `Server_auth)
-      Packet.BAD_CERTIFICATE >|= fun () ->
-    ((s, xs), Some anchor)
+    authenticate authenticator hostname certs >>= fun anchor ->
+    key_size Config.min_rsa_key_size certs >>= fun () ->
+    ( match certs with
+      | s :: _ ->
+        guard (validate_keytype s keytype &&
+               validate_usage s usage &&
+               validate_ext_usage s `Server_auth)
+          Packet.BAD_CERTIFICATE
+      | _ -> return () ) >|= fun () ->
```

Comment:
I've the feeling this `match certs with` and keytype and usage checks need to be moved to handshake_client, since the validation of client certs will most likely be different!? (currently for client auth, hostname is `None`, keytype `RSA` and usage `Digital_signature`)